### ISSUE TITLE
[SSH2] Add support for persistent connection

### ISSUE THREAD

**Issue by hason** [phpseclib/phpseclib#121]
I would like to add a configuration parameter for open persistent connection.

**Comment by oijkn** [comment]
Hi, sorry for this big bump but i need a persistent connection for my project.
The main command trough ssh run a deamon shell (moshell) and that take 15s. For now I have to reconnect every time and it's really long process. I tried to change pfsockopen() instead of just fsockopen() but the result is the same. There any alternative to maintain the ssh connection or call it back ? Thx you in advance for your support.

BR

**Comment by terrafrost** [comment]
Try something like this:

```php
<?php
include('Net/SSH2.php');

$ssh = new Net_SSH2('www.website.com');
$ssh->login('username', 'password');

$ssh->write("ping 127.0.0.1\n");

$start = microtime(true);
echo $ssh->exec('sleep 15s');
$elapsed = microtime(true) - $start;

echo "took $elapsed seconds\n";

echo $ssh->exec('ls -latr');
```

**Comment by oijkn** [comment]
@terrafrost This approach is good is you run a standard SSH connection. In my case I have another step after the SSH login, I execute a Daemon Shell (motshell for Motorola Management Device), this process need 10/15 sec to access at the app and then I execute a lot of successive commands. So you can see i can't use your "sleep" trick i would like to keep my motshell alive and access to it when i want. Can you help me with this request please ? Thx for the support !

BR

**Comment by terrafrost** [comment]
So I guess the specific issue with the code I posted is that you're already using the interactive read / write functions? In theory, something like this would work:

```php
<?php
include('Net/SSH2.php');

$ssh = new Net_SSH2('...');
if (!$ssh->login('user', 'pass')) {
	exit('Login Failed');
}

$ssh->write("ping 127.0.0.1\n");

$ssh->enablePTY();
$ssh->exec('bash');

$ssh->write("ls -latr\n");

$ssh->setTimeout(1);

echo $ssh->read();
```
In practice that doesn't work because a NET_SSH2_MSG_CHANNEL_DATA packet from the ping is being sent before the NET_SSH2_MSG_CHANNEL_SUCCESS from the bash. I'll try to get it so that can work but it'll be a few days before I'm able to work on it.

Right now read() / write() assume that you're talking about an interactive shell unless an exec with PTY was used at which point it assumes that that's being used. Once it makes that assumption you can't really go back.

Long term, for the current master branch, I'd like to make SSH2.php support proper multiple channels. All 65,536 possible channels (iirc that's how many channels SSH2 supports) but that's a very long term goal.

**Comment by oijkn** [comment]
@terrafrost Yes i use the interactive read/wrtie functions with enablePTY. Here is a piece of my code, for exemple : 

```php5
<?php
include('Net/SSH2.php');

$ssh = new Net_SSH2('...');
if (!$ssh->login('user', 'pass')) {
	exit('Login Failed');
}

$ssh->write("ping 127.0.0.1\n");

$ssh->enablePTY();
$ssh->read('/.*@.*[$|#|>]/', NET_SSH2_READ_REGEX);
$ssh->write("/usr/app/bin/motshell\n"); // daemon shell motorola
$ssh->setTimeout(20); // wait because motshell is very long to launch
$ssh->read('MotShell'); // motshell is ready
$ssh->write("mss get rf_msr mac " . $mac ."\n");
$ssh->read('MotShell');
//more code....
```

In my case I need to use the actual daemon shell **motshell** already running to send more commands later. What's your advice or trick to make it possible ?

Good luck for your project !!

Thank you for your support, best regards.
